local test = require("@std/test")
local cli = require("@batteries/cli")

test.suite("CLI flags", function(suite)
	suite:case("longFlag", function(assert)
		local p = cli.parser()
		p:add("verbose", "flag", {})
		p:parse({ "--verbose" })
		assert.that(p:has("verbose"))
	end)

	suite:case("shortFlag", function(assert)
		local p = cli.parser()
		p:add("verbose", "flag", { aliases = { "v" } })
		p:parse({ "-v" })
		assert.that(p:has("verbose"))
	end)

	suite:case("absentFlagIsFalse", function(assert)
		local p = cli.parser()
		p:add("verbose", "flag", {})
		p:parse({})
		assert.eq(p:has("verbose"), false)
	end)

	suite:case("combinedShortFlags", function(assert)
		local p = cli.parser()
		p:add("all", "flag", { aliases = { "a" } })
		p:add("long", "flag", { aliases = { "l" } })
		p:add("human", "flag", { aliases = { "h" } })
		p:parse({ "-alh" })
		assert.that(p:has("all"))
		assert.that(p:has("long"))
		assert.that(p:has("human"))
	end)

	suite:case("unknownLongFlagThrows", function(assert)
		local p = cli.parser()
		assert.throwsWith("Unknown argument: unknown", function()
			p:parse({ "--unknown" })
		end)
	end)

	suite:case("unknownShortFlagThrows", function(assert)
		local p = cli.parser()
		assert.throwsWith("Unknown argument: z", function()
			p:parse({ "-z" })
		end)
	end)
end)

test.suite("CLI options", function(suite)
	suite:case("longOption", function(assert)
		local p = cli.parser()
		p:add("output", "option", {})
		p:parse({ "--output", "file.txt" })
		assert.eq(p:get("output"), "file.txt")
	end)

	suite:case("shortOption", function(assert)
		local p = cli.parser()
		p:add("output", "option", { aliases = { "o" } })
		p:parse({ "-o", "file.txt" })
		assert.eq(p:get("output"), "file.txt")
	end)

	suite:case("missingOptionValueThrows", function(assert)
		local p = cli.parser()
		p:add("output", "option", {})
		assert.throwsWith("Missing value for argument: output", function()
			p:parse({ "--output" })
		end)
	end)

	suite:case("defaultAppliedWhenAbsent", function(assert)
		local p = cli.parser()
		p:add("output", "option", { default = "out.txt" })
		p:parse({})
		assert.eq(p:get("output"), "out.txt")
	end)

	suite:case("defaultNotAppliedWhenPresent", function(assert)
		local p = cli.parser()
		p:add("output", "option", { default = "out.txt" })
		p:parse({ "--output", "other.txt" })
		assert.eq(p:get("output"), "other.txt")
	end)

	suite:case("absentOptionReturnsNil", function(assert)
		local p = cli.parser()
		p:add("output", "option", {})
		p:parse({})
		assert.eq(p:get("output"), nil)
	end)

	suite:case("requiredOptionThrowsWhenAbsent", function(assert)
		local p = cli.parser()
		p:add("output", "option", { required = true })
		assert.throwsWith("Missing required argument: output", function()
			p:parse({})
		end)
	end)
end)

test.suite("CLI positional arguments", function(suite)
	suite:case("singlePositional", function(assert)
		local p = cli.parser()
		p:add("file", "positional", {})
		p:parse({ "foo.txt" })
		assert.eq(p:get("file"), "foo.txt")
	end)

	suite:case("multiplePositionals", function(assert)
		local p = cli.parser()
		p:add("src", "positional", {})
		p:add("dst", "positional", {})
		p:parse({ "a.txt", "b.txt" })
		assert.eq(p:get("src"), "a.txt")
		assert.eq(p:get("dst"), "b.txt")
	end)

	suite:case("positionalInterleavedWithFlags", function(assert)
		local p = cli.parser()
		p:add("verbose", "flag", {})
		p:add("file", "positional", {})
		p:parse({ "--verbose", "foo.txt" })
		assert.that(p:has("verbose"))
		assert.eq(p:get("file"), "foo.txt")
	end)

	suite:case("requiredPositionalThrowsWhenAbsent", function(assert)
		local p = cli.parser()
		p:add("file", "positional", { required = true })
		assert.throwsWith("Missing required argument: file", function()
			p:parse({})
		end)
	end)
end)

test.suite("CLI forwarded arguments", function(suite)
	suite:case("argsAfterDashDashAreForwarded", function(assert)
		local p = cli.parser()
		p:parse({ "--", "foo", "bar" })
		assert.eq(p:forwarded(), { "foo", "bar" })
	end)

	suite:case("dashDashWithFlagsBefore", function(assert)
		local p = cli.parser()
		p:add("verbose", "flag", {})
		p:parse({ "--verbose", "--", "--not-a-flag" })
		assert.that(p:has("verbose"))
		assert.eq(p:forwarded(), { "--not-a-flag" })
	end)

	suite:case("noForwardedArgsReturnsEmpty", function(assert)
		local p = cli.parser()
		p:parse({})
		assert.eq(p:forwarded(), {})
	end)

	suite:case("unconsumedPositionalsAreForwarded", function(assert)
		local p = cli.parser()
		p:parse({ "extra" })
		assert.eq(p:forwarded(), { "extra" })
	end)
end)

test.suite("CLI variadic arguments", function(suite)
	suite:case("capturesSingleArg", function(assert)
		local p = cli.parser()
		p:add("files", "variadic", {})
		p:parse({ "a.txt" })
		assert.eq(p:variadics("files"), { "a.txt" })
	end)

	suite:case("capturesMultipleArgs", function(assert)
		local p = cli.parser()
		p:add("files", "variadic", {})
		p:parse({ "a.txt", "b.txt", "c.txt" })
		assert.eq(p:variadics("files"), { "a.txt", "b.txt", "c.txt" })
	end)

	suite:case("noArgsReturnsEmpty", function(assert)
		local p = cli.parser()
		p:add("files", "variadic", {})
		p:parse({})
		assert.eq(p:variadics("files"), {})
	end)

	suite:case("flagsStillParsedAlongsideVariadic", function(assert)
		local p = cli.parser()
		p:add("verbose", "flag", {})
		p:add("files", "variadic", {})
		p:parse({ "--verbose", "a.txt", "b.txt" })
		assert.that(p:has("verbose"))
		assert.eq(p:variadics("files"), { "a.txt", "b.txt" })
	end)

	suite:case("optionsStillParsedAlongsideVariadic", function(assert)
		local p = cli.parser()
		p:add("output", "option", {})
		p:add("files", "variadic", {})
		p:parse({ "--output", "file.txt", "a.txt", "b.txt" })
		assert.eq(p:get("output"), "file.txt")
		assert.eq(p:variadics("files"), { "a.txt", "b.txt" })
	end)

	suite:case("positionalsConsumedBeforeVariadic", function(assert)
		local p = cli.parser()
		p:add("first", "positional", {})
		p:add("rest", "variadic", {})
		p:parse({ "one", "two", "three" })
		assert.eq(p:get("first"), "one")
		assert.eq(p:variadics("rest"), { "two", "three" })
	end)

	suite:case("argsAfterDashDashGoToForwardedNotVariadic", function(assert)
		local p = cli.parser()
		p:add("files", "variadic", {})
		p:parse({ "a.txt", "--", "b.txt" })
		assert.eq(p:variadics("files"), { "a.txt" })
		assert.eq(p:forwarded(), { "b.txt" })
	end)

	suite:case("requiredVariadicThrowsWhenEmpty", function(assert)
		local p = cli.parser()
		p:add("files", "variadic", { required = true })
		assert.throwsWith("Missing required variadic argument: files", function()
			p:parse({})
		end)
	end)

	suite:case("requiredVariadicPassesWhenPopulated", function(assert)
		local p = cli.parser()
		p:add("files", "variadic", { required = true })
		p:parse({ "a.txt" })
		assert.eq(p:variadics("files"), { "a.txt" })
	end)
end)

test.suite("CLI mixed usage", function(suite)
	suite:case("flagsOptionsAndPositionals", function(assert)
		local p = cli.parser()
		p:add("verbose", "flag", { aliases = { "v" } })
		p:add("output", "option", { aliases = { "o" } })
		p:add("input", "positional", {})
		p:parse({ "-v", "--output", "out.txt", "in.txt" })
		assert.that(p:has("verbose"))
		assert.eq(p:get("output"), "out.txt")
		assert.eq(p:get("input"), "in.txt")
	end)

	suite:case("positionalsThenForwardedAfterDashDash", function(assert)
		local p = cli.parser()
		p:add("file", "positional", {})
		p:parse({ "foo.txt", "--", "extra1", "extra2" })
		assert.eq(p:get("file"), "foo.txt")
		assert.eq(p:forwarded(), { "extra1", "extra2" })
	end)
end)
