local deque = require("@batteries/collections/deque")
local test = require("@std/test")

test.suite("deque", function(suite)
	suite:case("deque.new with no value creates empty deque", function(assert)
		local deq = deque.new()
		assert.eq(deq:peekfront(), nil)
		assert.eq(deq:peekback(), nil)
		assert.eq(#deq, 0)
	end)

	suite:case("deque.new with values creates deque with one element", function(assert)
		local deq = deque.new(1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 1)
		assert.eq(#deq, 1)
	end)

	suite:case("deque:pushfront", function(assert)
		local deq = deque.new(2)
		deq:pushfront(1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 2)
		assert.eq(#deq, 2)
	end)

	suite:case("deque:pushback", function(assert)
		local deq = deque.new(2)
		deq:pushback(1)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 1)
		assert.eq(#deq, 2)
	end)

	suite:case("deque:popfront", function(assert)
		local deq = deque.new(2)
		local popped = deq:popfront()
		assert.eq(popped, 2)
		assert.eq(#deq, 0)
	end)

	suite:case("deque:popback", function(assert)
		local deq = deque.new(2)
		local popped = deq:popback()
		assert.eq(popped, 2)
		assert.eq(#deq, 0)
	end)

	suite:case("deque:peekfront", function(assert)
		local deq = deque.new(1)
		assert.eq(#deq, 1)
		assert.eq(deq:peekfront(), 1)
	end)

	suite:case("deque:peekback", function(assert)
		local deq = deque.new(1)
		assert.eq(#deq, 1)
		assert.eq(deq:peekback(), 1)
	end)

	suite:case("popping from empty deque", function(assert)
		local deq = deque.new()
		assert.erroreq(function()
			deq:popback()
		end, "Popping from empty deque")
		assert.erroreq(function()
			deq:popfront()
		end, "Popping from empty deque")
	end)

	suite:case("e2e", function(assert)
		local deq = deque.new()
		deq:pushback(1)
		deq:pushback(2)
		deq:pushback(3)

		assert.eq(#deq, 3)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 3)

		assert.eq(deq:popfront(), 1)
		assert.eq(#deq, 2)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 3)

		assert.eq(deq:popback(), 3)
		assert.eq(#deq, 1)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 2)
	end)
end)

test.run()
