local deque = require("@batteries/collections/deque")
local test = require("@std/test")

test.suite("deque", function(suite)
	suite:case("deque.new with no value creates empty deque", function(assert)
		local deq = deque.new()
		assert.eq(deq:peekfront(), nil)
		assert.eq(deq:peekback(), nil)
		assert.eq(deq.len, 0)
	end)

	suite:case("deque.new with values creates deque with one element", function(assert)
		local deq = deque.new(1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 1)
		assert.eq(deq.len, 1)
	end)

	suite:case("deque:pushfront", function(assert)
		local deq = deque.new(2)
		deq:pushfront(1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 2)
		assert.eq(deq.len, 2)
	end)

	suite:case("deque:pushback", function(assert)
		local deq = deque.new(2)
		deq:pushback(1)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 1)
		assert.eq(deq.len, 2)
	end)

	suite:case("deque:popfront", function(assert)
		local deq = deque.new(2)
		local popped = deq:popfront()
		assert.eq(popped, 2)
		assert.eq(deq.len, 0)
	end)

	suite:case("deque:popback", function(assert)
		local deq = deque.new(2)
		local popped = deq:popback()
		assert.eq(popped, 2)
		assert.eq(deq.len, 0)
	end)

	suite:case("popping from size 1 -> 0", function(assert)
		local deq = deque.new(1)
		assert.eq(deq.len, 1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 1)
		deq:popback()
		assert.eq(deq.len, 0)
		assert.eq(deq:peekfront(), nil)
		assert.eq(deq:peekback(), nil)

		local deq2 = deque.new(1)
		assert.eq(deq2.len, 1)
		assert.eq(deq2:peekfront(), 1)
		deq2:popfront()
		assert.eq(deq2.len, 0)
		assert.eq(deq2:peekfront(), nil)
		assert.eq(deq2:peekback(), nil)
	end)

	suite:case("pushing from size 0 -> 1", function(assert)
		local deq = deque.new()
		assert.eq(deq.len, 0)
		assert.eq(deq:peekfront(), nil)
		assert.eq(deq:peekback(), nil)
		deq:pushfront(1)
		assert.eq(deq.len, 1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 1)

		deq = deque.new()
		assert.eq(deq.len, 0)
		assert.eq(deq:peekfront(), nil)
		assert.eq(deq:peekback(), nil)
		deq:pushback(1)
		assert.eq(deq.len, 1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 1)
	end)

	suite:case("popping from empty queue", function(assert)
		local deq = deque.new()
		assert.errors(function()
			deq:popback()
		end)
		assert.errors(function()
			deq:popfront()
		end)
	end)

	suite:case("e2e sanity check", function(assert)
		local deq = deque.new()
		deq:pushback(1)
		deq:pushback(2)
		deq:pushback(3)

		assert.eq(deq.len, 3)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 3)

		deq:popfront()
		assert.eq(deq.len, 2)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 3)

		deq:popback()
		assert.eq(deq.len, 1)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 2)
	end)
end)

test.run()
