local deque = require("@batteries/deque")
local test = require("@std/test")

test.suite("deque", function(suite)
	suite:case("deque.new creates empty deque", function(assert)
		local deq = deque.new()
		assert.eq(deq.head, nil)
		assert.eq(deq.tail, nil)
		assert.eq(deq.len, 0)
	end)

	suite:case("deque:pushfront", function(assert)
		local deq = deque.new(2)
		assert.eq((deq.head :: deque.DequeNode<number>).value, 2)
		assert.eq((deq.tail :: deque.DequeNode<number>).value, 2)

		deq:pushfront(1)
		assert.eq((deq.head :: deque.DequeNode<number>).value, 1)
		assert.eq((deq.tail :: deque.DequeNode<number>).value, 2) -- tail is unchanged
		assert.eq((deq.head :: deque.DequeNode<number>).next, deq.tail)
		assert.eq((deq.tail :: deque.DequeNode<number>).prev, deq.head)
		assert.eq(deq.len, 2)
	end)

	suite:case("deque:pushback", function(assert)
		local deq = deque.new(2)
		assert.eq((deq.head :: deque.DequeNode<number>).value, 2)
		assert.eq((deq.tail :: deque.DequeNode<number>).value, 2)

		deq:pushback(1)
		assert.eq((deq.head :: deque.DequeNode<number>).value, 2)
		assert.eq((deq.tail :: deque.DequeNode<number>).value, 1)
		assert.eq((deq.head :: deque.DequeNode<number>).next, deq.tail)
		assert.eq((deq.head :: deque.DequeNode<number>).prev, nil)
		assert.eq((deq.tail :: deque.DequeNode<number>).prev, deq.head)
		assert.eq((deq.tail :: deque.DequeNode<number>).next, nil)
		assert.eq(deq.len, 2)
	end)

	suite:case("deque:popfront", function(assert)
		local deq = deque.new(2)
		deq:pushback(1)
		local prepopTail = deq.tail
		local popped = deq:popfront()
		assert.eq(popped, 2)
		assert.eq(deq.len, 1)
		assert.eq(deq.head, prepopTail)
		assert.eq(deq.tail, prepopTail)
		assert.eq(prepopTail.next, nil)
		assert.eq(prepopTail.prev, nil)
	end)

	suite:case("deque:popback", function(assert)
		local deq = deque.new(2)
		deq:pushback(1)
		local prepopHead = deq.head
		local popped = deq:popback()
		assert.eq(popped, 1)
		assert.eq(deq.len, 1)
		assert.eq(deq.head, prepopHead)
		assert.eq(deq.tail, prepopHead)
		assert.eq(prepopHead.next, nil)
		assert.eq(prepopHead.prev, nil)
	end)

	suite:case("popping from size 1 -> 0", function(assert)
		local deq = deque.new(1)
		assert.neq(deq.head, nil)
		assert.neq(deq.tail, nil)
		deq:popback()
		assert.eq(deq.head, nil)
		assert.eq(deq.tail, nil)

		local deq2 = deque.new(1)
		assert.neq(deq2.head, nil)
		assert.neq(deq2.tail, nil)
		deq2:popfront()
		assert.eq(deq2.head, nil)
		assert.eq(deq2.tail, nil)
	end)

	suite:case("pushing from size 0 -> 1", function(assert)
		local deq = deque.new()
		assert.eq(deq.head, nil)
		assert.eq(deq.tail, nil)
		deq:pushfront(1)
		assert.neq(deq.head, nil)
		assert.neq(deq.tail, nil)
		assert.eq(deq.head, deq.tail)
		assert.eq((deq.head :: deque.DequeNode<number>).next, nil)
		assert.eq((deq.head :: deque.DequeNode<number>).prev, nil)

		deq = deque.new()
		assert.eq(deq.head, nil)
		assert.eq(deq.tail, nil)
		deq:pushback(1)
		assert.neq(deq.head, nil)
		assert.neq(deq.tail, nil)
		assert.eq(deq.head, deq.tail)
		assert.eq((deq.head :: deque.DequeNode<number>).next, nil)
		assert.eq((deq.head :: deque.DequeNode<number>).prev, nil)
	end)

	suite:case("popping from empty queue", function(assert)
		local deq = deque.new()
		assert.errors(function()
			deq:popback()
		end)
		assert.errors(function()
			deq:popfront()
		end)
	end)

	suite:case("e2e sanity check", function(assert)
		local deq = deque.new()
		deq:pushback(1)
		deq:pushback(2)
		deq:pushback(3)

		assert.eq(deq.len, 3)
		assert.eq(deq.head.value, 1)

		assert.eq(deq.head.next.value, 2)
		assert.eq(deq.head.next.prev.value, 1)

		assert.eq(deq.head.next.next.value, 3)
		assert.eq(deq.head.next.next.prev.value, 2)

		deq:popfront()
		deq:popback()
		assert.eq(deq.len, 1)
		assert.eq(deq.head.value, 2)
		assert.eq(deq.head.next, nil)
		assert.eq(deq.head.prev, nil)
		assert.eq(deq.head, deq.tail)
	end)
end)

test.run()
