local deque = require("@batteries/collections/deque")
local test = require("@std/test")

test.suite("Deque", function(suite)
	suite:case("deque_new_with_no_value_creates_empty_deque", function(assert)
		local deq = deque.new()
		assert.eq(deq:peekfront(), nil)
		assert.eq(deq:peekback(), nil)
		assert.eq(#deq, 0)
	end)

	suite:case("deque_new_with_values_creates_deque_with_one_element", function(assert)
		local deq = deque.new(1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 1)
		assert.eq(#deq, 1)
	end)

	suite:case("deque_pushfront", function(assert)
		local deq = deque.new(2)
		deq:pushfront(1)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 2)
		assert.eq(#deq, 2)
	end)

	suite:case("deque_pushback", function(assert)
		local deq = deque.new(2)
		deq:pushback(1)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 1)
		assert.eq(#deq, 2)
	end)

	suite:case("deque_popfront", function(assert)
		local deq = deque.new(2)
		local popped: number = deq:popfront()
		assert.eq(popped, 2)
		assert.eq(#deq, 0)
		assert.eq(deq:peekfront(), nil)
		assert.eq(deq:peekback(), nil)
	end)

	suite:case("deque_popback", function(assert)
		local deq = deque.new(2)
		local popped: number = deq:popback()
		assert.eq(popped, 2)
		assert.eq(#deq, 0)
		assert.eq(deq:peekfront(), nil)
		assert.eq(deq:peekback(), nil)
	end)

	suite:case("deque_peekfront", function(assert)
		local deq = deque.new(1)
		assert.eq(#deq, 1)
		assert.eq(deq:peekfront(), 1)
	end)

	suite:case("deque_peekback", function(assert)
		local deq = deque.new(1)
		assert.eq(#deq, 1)
		assert.eq(deq:peekback(), 1)
	end)

	suite:case("popping_from_empty_deque", function(assert)
		local deq = deque.new()
		assert.erroreq(function()
			deq:popback()
		end, "Popping from empty deque")
		assert.erroreq(function()
			deq:popfront()
		end, "Popping from empty deque")
	end)

	suite:case("e2e", function(assert)
		local deq = deque.new()
		deq:pushback(1)
		deq:pushback(2)
		deq:pushback(3)

		assert.eq(#deq, 3)
		assert.eq(deq:peekfront(), 1)
		assert.eq(deq:peekback(), 3)

		assert.eq(deq:popfront(), 1)
		assert.eq(#deq, 2)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 3)

		assert.eq(deq:popback(), 3)
		assert.eq(#deq, 1)
		assert.eq(deq:peekfront(), 2)
		assert.eq(deq:peekback(), 2)
	end)
end)
